pkg/rebalancer/rebalancer.go

```diff
@@ -0,0 +1,305 @@
+		return false, nil
+	}
+
+	for _, rtd := range resourcesToDelete {
+		klog.V(2).Infof("Rebalancing pod %s/%s.", rtd.pod.Namespace, rtd.pod.Name)
+		err := r.rebalanceWorkload(ctx, rtd.pod, rtd.pvc)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// If the workloads were balanced by the operator, we wait for 5 minutes
+	// before setting the status so that we don't set upgradeable=false after
+	// balancing the pods.
+	err = wait.Poll(10*time.Second, 5*time.Minute, func() (bool, error) {
+		klog.V(4).Infof("Waiting until workload in namespace %s with label %q becomes correctly balanced.", workload.Namespace, workload.LabelSelector)
+		return r.WorkloadCorrectlyBalanced(ctx, workload)
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return true, r.ensurePVCsAreNotAnnotated(ctx, workload)
+}
+
+// rebalanceWorkload rebalances the given pod on a different node. To make sure
+// that the pod will be rescheduled on a different node, this function will
+// first cordon the node in which the pod is scheduled. Then it will verify
+// whether the PVC attached to the pod needs to be deleted or not. Only pods
+// that have PVCs with a zonal topology need to be deleted since they might
+// prevent the pod to be schedule on a node that is located in a different
+// availability zone.
+func (r *Rebalancer) rebalanceWorkload(ctx context.Context, pod *v1.Pod, pvc *v1.PersistentVolumeClaim) error {
+	node, err := r.client.CoreV1().Nodes().Get(ctx, pod.Spec.NodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = drain.RunCordonOrUncordon(r.drainer, node, drain.Cordon)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := drain.RunCordonOrUncordon(r.drainer, node, drain.Uncordon)
+		if err != nil {
+			klog.Errorf("Couldn't uncordon node %v: %v.", pod.Spec.NodeName, err)
+		}
+	}()
+
+	pv, err := r.client.CoreV1().PersistentVolumes().Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// Do not delete the PVC if the storage provider hasn't set the topology.kubernetes.io/zone label on the PV.
+	// In most cases, when the PV isn't zonal, pods can access it from a node in a different AZ so we don't need to delete it.
+	if _, ok := pv.GetLabels()[zonalTopologyAnnotation]; ok {
+		klog.V(2).Infof("Deleting PersistentVolumeClaim %s/%s.", pvc.Namespace, pvc.Name)
+		err = r.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		klog.V(2).Infof("Keeping PersistentVolumeClaim %s/%s because it doesn't have the %q label", pvc.Namespace, pvc.Name, zonalTopologyAnnotation)
+	}
+
+	klog.V(2).Infof("Deleting pod %s/%s.", pod.Namespace, pod.Name)
+	return r.client.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+}
+
+type resourcesToDelete struct {
+	pod *v1.Pod
+	pvc *v1.PersistentVolumeClaim
+}
+
+type byAge []resourcesToDelete
```

Comment:
```suggestion
// byAge sorts resources by PVC creation time in reverse chronological order (most recent first).
// In case of a tie it sorts the resources by the
// pod names in reverse alphabetical order (e.g. [pod-1, pod-0]).
type byAge []resourcesToDelete
```